force_nevergrad/engine/nevergrad_optimizers.py

```diff
@@ -0,0 +1,215 @@
+import numpy as np
+from functools import partial
+
+from traits.api import (
+    Enum,
+    provides,
+    HasStrictTraits,
+    List
+)
+
+from force_bdss.api import (
+    PositiveInt,
+    KPISpecification,
+    IOptimizer
+)
+
+from .parameter_translation import (
+    translate_mco_to_ng,
+    translate_ng_to_mco
+)
+import nevergrad as ng
+from nevergrad.functions import MultiobjectiveFunction
+
+
+def nevergrad_function(*ng_params,
+                       function=None,
+                       is_scalar=True,
+                       minimize_objectives=[]):
+    """ A wrapper around the MCO objective function,
+    that can be optimized by nevergrad.
+
+    Parameters
+    ----------
+    *ng_params: Any (but usually float, ndarray or string)
+        Parameters to be optimized.
+    function: Callable
+        The MCO objective function (_score(), etc).
+    is_scalar: bool
+        Whether or not the function should be scalar.
+        (be a single-objective function).
+    minimize_objectives: list of bool
+        For each objective, whether it should be minimized
+        (as opposed to maximized).
+
+    Return
+    ------
+    float or ndarray
+        The objectives/kpis or their sum.
+
+    Notes
+    -----
+    Nevergrad must optimize a function where each positional or
+    keyword argument is a single parameter.
+    foo(p1, p2, p3, ...pN)
+    MCO objective functions (_score(), etc.) have a single argument -
+    the list of parameters.
+    foo([p1, p2, p3, ...pN])
+    The types of the individual parameters are almost the same, in the
+    nevergrad and MCO functions (but see translate_ng_to_mco()).
+    Both nevergrad and MCO functions return the objectives/kpis as
+    a single, one-dimensional numpy array.
+    """
+
+    # pack the nevergrad parameters values
+    # into a list of mco parameter values.
+    mco_params = translate_ng_to_mco(list(ng_params))
+
+    # call the MCO objective function
+    objective = function(mco_params)
+
+    # negate any objectives that should be maximised
+    if len(minimize_objectives) == len(objective):
+        for i, do_minimize in enumerate(minimize_objectives):
+            if not do_minimize:
+                objective[i] = -objective[i]
+
+    # if the objective needs to be scalar but it is not,
+    # sum the objectives into a single objective.
+    if is_scalar and not np.isscalar(objective):
+        return np.sum(objective)
+    return objective
+
+
+@provides(IOptimizer)
+class NevergradScalarOptimizer(HasStrictTraits):
+    """ Optimization of a scalar function using nevergrad.
+    """
+
+    #: Algorithms available to work with
+    algorithms = Enum(*ng.optimizers.registry.keys())
```

Comment:
We might want to define this list of keys as a single object outside the class definitions as it is being used in a few different places.

```python
ALGORITHMS_KEYS = ng.optimizers.registry.keys()
...
    algorithms = Enum(*ALGORITHMS_KEYS)
```

Then the `NevergradMCOModel` can also import this object directly, rather than using the `*NevergradMultiOptimizer.class_traits()["algorithms"].handler.values` that was put in previously.